adapta/storage/blob/s3_storage_client.py

```diff
@@ -0,0 +1,145 @@
+"""
+ Storage Client implementation for AWS S3 Cloud.
+"""
+#  Copyright (c) 2023. ECCO Sneaks & Data
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+
+from abc import ABC
+from typing import Optional, Callable, Type, Iterator, Dict, TypeVar
+
+from adapta.security.clients import AwsClient
+from adapta.storage.blob.base import StorageClient
+from adapta.storage.models.aws import cast_path
+from adapta.storage.models.base import DataPath
+from adapta.storage.models.format import SerializationFormat
+
+T = TypeVar("T")  # pylint: disable=C0103
+
+
+class S3StorageClient(StorageClient, ABC):
+    """
+    S3 Storage Client.
+    """
+
+    def __init__(self, *, base_client: AwsClient, bucket_name: str):
```

Comment:
You have an issue with this, as people can pass DataPaths targeting other buckets. I think you should instead have s3 resource and create bucket mapping on the fly similar to what we do with Azure